File: python/agentpin/nonce.py
"""Nonce deduplication for replay attack prevention."""

import threading
import time


class NonceStore:
    """Abstract base for nonce deduplication."""

    def check_and_record(self, nonce: str, ttl_seconds: float) -> bool:
        """Check if nonce is fresh. Returns True if fresh, False if replay."""
        raise NotImplementedError


class InMemoryNonceStore(NonceStore):
    """In-memory nonce store with lazy expiry cleanup."""

    def __init__(self):
        self._entries = {}  # nonce -> expiry_time
        self._lock = threading.Lock()

    def check_and_record(self, nonce: str, ttl_seconds: float) -> bool:
        with self._lock:
            now = time.monotonic()
            # Lazy cleanup
            self._entries = {k: v for k, v in self._entries.items() if v > now}
            # Check
            if nonce in self._entries:
                return False
            # Record
            self._entries[nonce] = now + ttl_seconds
            return True

File: javascript/src/nonce.js
/**
 * Nonce deduplication for replay attack prevention.
 */

/**
 * In-memory nonce store that tracks seen nonces with TTL-based expiry.
 */
export class InMemoryNonceStore {
    constructor() {
        /** @type {Map<string, number>} nonce -> expiry timestamp (ms) */
        this._entries = new Map();
    }

    /**
     * Check if nonce is fresh. Returns true if fresh, false if replay.
     * @param {string} nonce
     * @param {number} ttlMs - TTL in milliseconds
     * @returns {boolean}
     */
    checkAndRecord(nonce, ttlMs) {
        const now = Date.now();
        // Lazy cleanup
        for (const [key, expiry] of this._entries) {
            if (expiry <= now) this._entries.delete(key);
        }
        if (this._entries.has(nonce)) return false;
        this._entries.set(nonce, now + ttlMs);
        return true;
    }
}
